packages/taquito/src/prepare/prepare-provider.ts

```diff
@@ -0,0 +1,863 @@
+  VotingPeriodBlockResult,
+} from '@taquito/rpc';
+import {
+  DelegateParams,
+  RevealParams,
+  RPCOperation,
+  RPCOpWithFee,
+  RPCOpWithSource,
+  isOpRequireReveal,
+  RegisterGlobalConstantParams,
+  TransferParams,
+  OriginateParams,
+  TxRollupOriginateParams,
+  TxRollupBatchParams,
+  UpdateConsensusKeyParams,
+  TransferTicketParams,
+  IncreasePaidStorageParams,
+  BallotParams,
+  ProposalsParams,
+  DrainDelegateParams,
+  ParamsWithKind,
+} from '../operations/types';
+import { PreparationProvider, PreparedOperation } from './interface';
+import { Protocols } from '../constants';
+import { InvalidOperationKindError, DeprecationError } from '@taquito/utils';
+import { RPCResponseError } from '../error';
+import { Context } from '../context';
+import { ContractMethod } from '../contract/contract-methods/contract-method-flat-param';
+import { ContractMethodObject } from '../contract/contract-methods/contract-method-object-param';
+import { ContractProvider } from '../contract/interface';
+import {
+  createSetDelegateOperation,
+  createTransferOperation,
+  createRevealOperation,
+  createRegisterGlobalConstantOperation,
+  createOriginationOperation,
+  RevealOperationError,
+  createTxRollupOriginationOperation,
+  createTxRollupBatchOperation,
+  createUpdateConsensusKeyOperation,
+  createTransferTicketOperation,
+  createIncreasePaidStorageOperation,
+  createBallotOperation,
+  createProposalsOperation,
+  createDrainDelegateOperation,
+  DefaultContractType,
+  ContractStorageType,
+} from '../contract';
+import { Estimate } from '../estimate';
+import { OperationBatch } from '../batch/rpc-batch-provider';
+
+/**
+ * @description PrepareProvider is a utility class to output the prepared format of an operation
+ */
+export class PrepareProvider implements PreparationProvider {
+  #counters: { [key: string]: number };
+
+  constructor(protected context: Context) {
+    this.#counters = {};
+  }
+
+  get rpc(): RpcClientInterface {
+    return this.context.rpc;
+  }
+
+  get signer() {
+    return this.context.signer;
+  }
+
+  get estimate() {
+    return this.context.estimate;
+  }
+
+  get parser() {
+    return this.context.parser;
+  }
+
+  private async getPkh() {
+    return await this.context.signer.publicKeyHash();
+  }
+
+  private async getBlockHash() {
+    return this.context.readProvider.getBlockHash('head~2');
+  }
+
+  private async getProtocolHash() {
+    return this.context.readProvider.getNextProtocol('head');
+  }
+
+  private async getHeadCounter(pkh: string): Promise<string> {
+    return this.context.readProvider.getCounter(pkh, 'head') ?? '0';
+  }
+
+  private getFee(op: RPCOpWithFee, pkh: string, headCounter: number) {
+    if (!this.#counters[pkh] || this.#counters[pkh] < headCounter) {
+      this.#counters[pkh] = headCounter;
+    }
+    const opCounter = ++this.#counters[pkh];
+
+    return {
+      counter: `${opCounter}`,
+      fee: typeof op.fee === 'undefined' ? '0' : `${op.fee}`,
```

Comment:
could any of these be something other than undefined? the type is `fee?: string` correct?
if so we could replace all of the `typeof op.fee <ternary op>` to eg: `fee: op.fee ?? '0'` and seems a little more readable to me 
what do you think? 
its a typescript nuance